pkg/auth/signatures.go

```diff
@@ -0,0 +1,138 @@
+package auth
+
+import (
+	"io"
+	"os"
+
+	"github.com/square/p2/pkg/util"
+	"golang.org/x/crypto/openpgp"
+)
+
+// Access to another goroutine that can verify PGP signatures. Close
+// the channel to cause the backing goroutine to exit.
+type SigVerifier chan<- SigRequest
+
+// CheckDetachedSignatureAsync() is a shortcut method to pack the
+// verification request into a struct and send it.
+func (v SigVerifier) CheckDetachedSignatureAsync(
+	signed, signature io.Reader,
+) <-chan SigReply {
+	rc := make(chan SigReply, 1)
+	v <- SigRequest{signed, signature, rc}
+	return rc
+}
+
+type SigRequest struct {
+	Signed, Signature io.Reader
+	ReplyChan         chan<- SigReply
+}
+
+type SigReply struct {
+	Entity *openpgp.Entity
+	Err    error
+}
+
+// Unpack() returns the reply as a canonical multi-valued return.
+func (r SigReply) Unpack() (*openpgp.Entity, error) {
+	return r.Entity, r.Err
+}
+
+// CheckDetachedSignature() performs a blocking call to the verifier.
+func (v SigVerifier) CheckDetachedSignature(
+	signed, signature io.Reader,
+) (*openpgp.Entity, error) {
+	return (<-v.CheckDetachedSignatureAsync(signed, signature)).Unpack()
+}
+
+// FuncVerifier() constructs a verifier that just calls a function to
+// verify the signature.
+func FuncVerifier(
+	verify func(signed, signature io.Reader) (*openpgp.Entity, error),
+) SigVerifier {
+	reqs := make(chan SigRequest)
+	go func() {
+		for req := range reqs {
+			entity, err := verify(req.Signed, req.Signature)
+			req.ReplyChan <- SigReply{entity, err}
+		}
+	}()
+	return SigVerifier(reqs)
+}
+
+// FixedKeyringVerifier() constructs a verifier that uses a fixed set of keys
+func FixedKeyringVerifier(entities ...*openpgp.Entity) SigVerifier {
+	keyring := openpgp.EntityList(entities)
+	return FuncVerifier(func(signed, signature io.Reader) (*openpgp.Entity, error) {
+		return openpgp.CheckDetachedSignature(keyring, signed, signature)
+	})
+}
+
+// FileKeyring() constructs a verifier that loads a keyring from a
+// file. The keyring will be reloaded whenever the file's modification
+// time changes.
+func FileKeyringVerifier(filename string) (SigVerifier, error) {
+	// Fail fast if it's an invalid file
+	info, err := os.Stat(filename)
+	if err != nil {
+		return nil, err
+	}
+	keyring, err := loadKeyring(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	reqs := make(chan SigRequest)
+	go func() {
+		for req := range reqs {
+			newInfo, err := os.Stat(filename)
+			if err != nil {
+				req.ReplyChan <- SigReply{nil, err}
+				continue
+			}
+			if info.ModTime() != newInfo.ModTime() {
+				// File has been modified! Reload it.
+				newKeyring, err := loadKeyring(filename)
+				if err != nil {
+					req.ReplyChan <- SigReply{nil, err}
```

Comment:
If I understand the overall system, the file that will be monitored with this is only going to be periodically changed by FCM, right? If that fails, it won't change the existing files on disk.

If you're thinking of an online method of collecting these keys, that can be supported, but it's a different `SigVerifier` instance. I think my preferred solution would be a layered approach:
1. In-memory keyring holding at least the keys to verify the currently-running pods
2. Disk cache of recent keys found
3. Master keyring distributed via FCM (to ensure that there is always someone who can deploy)
4. Online retriever that can retrieve keys as it needs when it tries to verify a signature
That and decoupling the possession of a key with authentication and authorization.